File: BFS/178.Graph Valid Tree/Solution_BFS.py
from collections import deque, defaultdict

class Solution:
    """
    @param n: An integer
    @param edges: a list of undirected edges
    @return: true if it's a valid tree, or false
    """
    def validTree(self, n, edges):
        # write your code here
        if n <= 1:
            return True

        # Contraint 1: # of edges equals to # of node minus 1
        if len(edges) != n - 1:
            return False

        # map from integer -> set of neighbors
        mapping = self._get_neighbors(edges)

        # Constraint 2: # of onnected component in the graph is 1
        queue = deque([])
        queue.append(0)
        history = set()

        while queue:
            node = queue.popleft()
            for neighbor in mapping[node]:
                if neighbor not in history:
                    queue.append(neighbor)
                    history.add(neighbor)

        if len(history) == n:
            return True
        return False

    def _get_neighbors(self, edges):
        mapping = defaultdict(set)
        for edge in edges:
            source, target = edge
            mapping[source].add(target)
            mapping[target].add(source)

        return mapping

File: BFS/178.Graph Valid Tree/Solution_BFS.java
public class Solution {
    /**
     * @param n: An integer
     * @param edges: a list of undirected edges
     * @return: true if it's a valid tree, or false
     */
    public boolean validTree(int n, int[][] edges) {
        // write your code here
        if (n == 0 || edges.length != n - 1) {
            return false;
        }
        
        // Collect neighbors
        Map<Integer, Set<Integer>> graph = initializeGraph(n, edges);

        // BFS
        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> history = new HashSet<>();

        // labeled from 0 to n - 1
        queue.offer(0);
        history.add(0);
        int num = 0;

        while (!queue.isEmpty()) {

            int node = queue.poll();
            num++;

            for (Integer neighbor : graph.get(node)) {
                if (!history.contains(neighbor)) {
                    history.add(neighbor);
                    queue.offer(neighbor);
                }
            }

        }

        return (num == n);
    }

    private Map<Integer, Set<Integer>> initializeGraph(int n, int[][]edges) {
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        // Initialization
        for (int i = 0; i < n; i++) {
            graph.put(i, new HashSet<Integer>());
        }
        
        // Add neighbors
        for (int[] pair : edges) {
            int node1 = pair[0];
            int node2 = pair[1];
            graph.get(node1).add(node2);
            graph.get(node2).add(node1);
        }

        return graph;
    }

}